flink-runtime/src/main/java/org/apache/flink/streaming/api/operators/StreamTaskStateInitializerImpl.java

```diff
@@ -252,31 +263,32 @@ public StreamOperatorStateContext streamOperatorStateContext(
                     (prioritizedOperatorSubtaskStates.isRestored() && !isUsingCustomRawKeyedState)
                             ? rawKeyedStateInputs
                             : Collections.emptyList();
-            if (keyedStatedBackend != null) {
-                timeServiceManager =
-                        timeServiceManagerProvider.create(
-                                environment.getMetricGroup().getIOMetricGroup(),
-                                keyedStatedBackend,
-                                keyedStatedBackend.getKeyGroupRange(),
-                                environment.getUserCodeClassLoader().asClassLoader(),
-                                keyContext,
-                                processingTimeService,
-                                restoredRawKeyedStateTimers,
-                                cancellationContext);
-            }
-            if (stateBackend.supportsAsyncKeyedStateBackend()) {
-                asyncTimeServiceManager =
-                        timeServiceManagerProvider.create(
-                                environment.getMetricGroup().getIOMetricGroup(),
-                                asyncKeyedStateBackend,
-                                asyncKeyedStateBackend.getKeyGroupRange(),
-                                environment.getUserCodeClassLoader().asClassLoader(),
-                                keyContext,
-                                processingTimeService,
-                                restoredRawKeyedStateTimers,
-                                cancellationContext);
+            if (isAsyncState) {
+                if (asyncKeyedStateBackend != null) {
```

Comment:
If `asyncKeyedStateBackend` is instance of `adaptor`, is it better to use the delegated `keyedStateBackend` to create `timerManager`?

Reply:
In this case, the `AsyncKeyedStateBackendAdaptor#create` is invoked. So yes, this is the current behavior.